src/System.Private.Xml/src/System/Xml/Core/XmlRawTextWriter.cs

```diff
@@ -0,0 +1,1623 @@
+            omitXmlDeclaration = settings.OmitXmlDeclaration;
+            newLineChars = settings.NewLineChars;
+            checkCharacters = settings.CheckCharacters;
+            closeOutput = settings.CloseOutput;
+
+            standalone = settings.Standalone;
+            outputMethod = settings.OutputMethod;
+            mergeCDataSections = settings.MergeCDataSections;
+
+            if (checkCharacters && newLineHandling == NewLineHandling.Replace)
+            {
+                ValidateContentChars(newLineChars, "NewLineChars", false);
+            }
+        }
+
+        // Returns settings the writer currently applies.
+        public override XmlWriterSettings Settings => GetSettings();
+
+        private XmlWriterSettings GetSettings()
+        {
+            return new XmlWriterSettings
+            {
+                Encoding = encoding,
+                OmitXmlDeclaration = omitXmlDeclaration,
+                NewLineHandling = newLineHandling,
+                NewLineChars = newLineChars,
+                CloseOutput = closeOutput,
+                ConformanceLevel = ConformanceLevel.Auto,
+                CheckCharacters = checkCharacters,
+                AutoXmlDeclaration = autoXmlDeclaration,
+                Standalone = standalone,
+                OutputMethod = outputMethod,
+                ReadOnly = true
+            };
+        }
+
+        protected XmlWriterSettings GetSettingsIndent()
+        {
+            var settings = GetSettings();
+
+            settings.ReadOnly = false;
+            settings.Indent = true;
+            settings.IndentChars = _indentChars;
+            settings.NewLineOnAttributes = _newLineOnAttributes;
+            settings.ReadOnly = true;
+
+            return settings;
+        }
+
+        // Write the xml declaration.  This must be the first call.
+        internal override void WriteXmlDeclaration(XmlStandalone standalone)
+        {
+            // Output xml declaration only if user allows it and it was not already output
+            if (!omitXmlDeclaration && !autoXmlDeclaration)
+            {
+                SetTextContentMark(false);
+
+                RawText("<?xml version=\"");
+
+                // Version
+                RawText("1.0");
+
+                // Encoding
+                if (encoding != null)
+                {
+                    RawText("\" encoding=\"");
+                    RawText(encoding.WebName);
+                }
+
+                // Standalone
+                if (standalone != XmlStandalone.Omit)
+                {
+                    RawText("\" standalone=\"");
+                    RawText(standalone == XmlStandalone.Yes ? "yes" : "no");
+                }
+
+                RawText("\"?>");
+            }
+        }
+
+        protected abstract bool WhileIsAttr(int ch);
+        protected abstract bool WhileNotSurHighStart(int ch);
+        protected abstract T ToBufferType(char ch);
```

Comment:
I would expect that these abstractions are going to introduce significant performance hit.

Reply:
My original profiling I base this off of (https://github.com/dotnet/corefx/issues/34200#issuecomment-449656139) sadly made a large oversight in it didn't account for it would be abstracted.

```CSharp
    abstract unsafe class BaseCastTest<T>
        where T : unmanaged
    {
        protected abstract T ToBufferType(char ch);

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        protected abstract T ToBufferTypeAggressive(char ch);

        protected T* buffer = (T*)Marshal.AllocHGlobal(32);

        public void IndirectlyCast()
        {
            for (var i = 0; i < Benchmark.InnerIterationCount; i++)
            {
                var bufPos = 0;
                buffer[bufPos++] = ToBufferType('"');
                buffer[bufPos++] = ToBufferType(' ');
                buffer[bufPos++] = ToBufferType('[');
                buffer[bufPos++] = ToBufferType(']');
                buffer[bufPos++] = ToBufferType('>');
                buffer[bufPos++] = ToBufferType('"');
                buffer[bufPos++] = ToBufferType('"');
                buffer[bufPos++] = ToBufferType(' ');
                buffer[bufPos++] = ToBufferType('[');
                buffer[bufPos++] = ToBufferType(']');
                buffer[bufPos++] = ToBufferType('>');
                buffer[bufPos++] = ToBufferType('"');
            }
        }

        public void IndirectlyCastAggressive()
        {
            for (var i = 0; i < Benchmark.InnerIterationCount; i++)
            {
                var bufPos = 0;
                buffer[bufPos++] = ToBufferTypeAggressive('"');
                buffer[bufPos++] = ToBufferTypeAggressive(' ');
                buffer[bufPos++] = ToBufferTypeAggressive('[');
                buffer[bufPos++] = ToBufferTypeAggressive(']');
                buffer[bufPos++] = ToBufferTypeAggressive('>');
                buffer[bufPos++] = ToBufferTypeAggressive('"');
                buffer[bufPos++] = ToBufferTypeAggressive('"');
                buffer[bufPos++] = ToBufferTypeAggressive(' ');
                buffer[bufPos++] = ToBufferTypeAggressive('[');
                buffer[bufPos++] = ToBufferTypeAggressive(']');
                buffer[bufPos++] = ToBufferTypeAggressive('>');
                buffer[bufPos++] = ToBufferTypeAggressive('"');
            }
        }
    }

    class CastTest : BaseCastTest<byte>
    {
        protected override byte ToBufferType(char ch) => (byte)ch;

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        protected override byte ToBufferTypeAggressive(char ch) => (byte)ch;

        public unsafe void DirectlyCast()
        {
            for (var i = 0; i < Benchmark.InnerIterationCount; i++)
            {
                var bufPos = 0;
                buffer[bufPos++] = (byte)'"';
                buffer[bufPos++] = (byte)'"';
                buffer[bufPos++] = (byte)' ';
                buffer[bufPos++] = (byte)'[';
                buffer[bufPos++] = (byte)']';
                buffer[bufPos++] = (byte)'>';
                buffer[bufPos++] = (byte)'"';
                buffer[bufPos++] = (byte)'"';
                buffer[bufPos++] = (byte)' ';
                buffer[bufPos++] = (byte)'[';
                buffer[bufPos++] = (byte)']';
                buffer[bufPos++] = (byte)'>';
            }
        }
    }

        [Benchmark(InnerIterationCount = 100000)]
        public void ProfileDirect()
        {
            var test = new CastTest();
            foreach (var iteration in Benchmark.Iterations)
            using (iteration.StartMeasurement())
            {
                test.DirectlyCast();
            }
        }

        [Benchmark(InnerIterationCount = 100000)]
        public void ProfileIndirectCasting()
        {
            var test = new CastTest();
            foreach (var iteration in Benchmark.Iterations)
            using (iteration.StartMeasurement())
            {
                test.IndirectlyCast();
            }
        }

        [Benchmark(InnerIterationCount = 100000)]
        public void ProfileIndirectCastingAggressive()
        {
            var test = new CastTest();
            foreach (var iteration in Benchmark.Iterations)
            using (iteration.StartMeasurement())
            {
                test.IndirectlyCastAggressive();
            }
        }
```

{noformat}                                 | Metric                                        | Unit  | Iterations |    Average |    STDEV.S |        Min |        Max
-------------------------------- |:--------------------------------------------- |:-----:|:----------:| ----------:| ----------:| ----------:| ----------:
ProfileDirect                    | Duration                                      | msec  |    100     |      0.324 |      0.040 |      0.288 |      0.500
ProfileDirect                    | Branch Mispredictions                         | count |    100     |    204.800 |    897.200 |      0.000 |   4096.000
ProfileDirect                    | Cache Misses                                  | count |    100     |    122.880 |    702.245 |      0.000 |   4096.000
ProfileDirect                    | Instructions Retired                          | count |    100     | 3.210E+006 | 409360.181 | 3.000E+006 | 4.000E+006
ProfileDirect                    | Allocation Size on Benchmark Execution Thread | bytes |    100     |      0.000 |      0.000 |      0.000 |      0.000
ProfileIndirectCasting           | Duration                                      | msec  |    100     |      1.639 |      0.133 |      1.490 |      2.168
ProfileIndirectCasting           | Branch Mispredictions                         | count |    100     |    819.200 |   1746.540 |      0.000 |   8192.000
ProfileIndirectCasting           | Cache Misses                                  | count |    100     |    286.720 |   1050.347 |      0.000 |   4096.000
ProfileIndirectCasting           | Instructions Retired                          | count |    100     | 1.449E+007 | 502418.394 | 1.400E+007 | 1.500E+007
ProfileIndirectCasting           | Allocation Size on Benchmark Execution Thread | bytes |    100     |      0.000 |      0.000 |      0.000 |      0.000
ProfileIndirectCastingAggressive | Duration                                      | msec  |    100     |      1.617 |      0.144 |      1.483 |      2.208
ProfileIndirectCastingAggressive | Branch Mispredictions                         | count |    100     |    696.320 |   1546.343 |      0.000 |   4096.000
ProfileIndirectCastingAggressive | Cache Misses                                  | count |    100     |    327.680 |   1116.815 |      0.000 |   4096.000
ProfileIndirectCastingAggressive | Instructions Retired                          | count |    100     | 1.447E+007 | 540388.936 | 1.300E+007 | 1.500E+007
ProfileIndirectCastingAggressive | Allocation Size on Benchmark Execution Thread | bytes |    100     |      0.000 |      0.000 |      0.000 |      0.000{noformat}

This is considerably more grim.

I have not experimented with it yet, but if I can mitigate the performance differences using `System.Linq.Expressions` would that be acceptable?

Reply:
You could try to seal your types, maybe the devirtualization code works here.
The functor will never inline, so it will always be slow.
Depending on the different types of T you'll need to support (e.g. char and byte) you could get rid of ``ToBufferType`` and directly implement your own and use ``Unsafe.As`` to get around the casting problems.

Reply:
Would it be acceptable for me to use the `Unsafe` class here? I've only just learned about it when researching this.

One alternative I came up with is `*(T*)&ch` which had an average execution time of 0.818. So twice the speed of the virtualized call, but still half as fast as otherwise.

Sadly I can't seal these classes as they're used elsewhere. By the html writer for example.

Reply:
> Would it be acceptable for me to use the Unsafe class here?

Unsafe may help you to reduce the overhead, but I doubt you would be able to fully recover the performance regression.

I think the best way to make this more maintainable are T4 templates. Generics do not work well for cases like these.

Reply:
Unless inlining of virtuals is improved we wont be able to mitigate the performance hit here. `ToBufferType` being moved to `Unsafe.As` would likely bring that into acceptable performance range, but I believe a real world benchmark would show `EncodeChar` is the main culprit, as it appears to do most of the buffer copying, and that's not such an easy fix.

I'll move this over to a t4 template. I'll also make a new template for the async ones.

I'd love to keep this large of a code reduction if possible. If anyone has ideas on how to make this acceptable let me know and I'll fix this up and open a new PR if it works out.